modules/3d/test/test_normal_estimation.cpp

```diff
@@ -0,0 +1,112 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2022, Wanli Zhong <zhongwl2018@mail.sustech.edu.cn>
+
+#include "test_precomp.hpp"
+#include "test_ptcloud_utils.hpp"
+
+namespace opencv_test { namespace {
+
+TEST(NormalEstimationTest, PlaneNormalEstimation)
+{
+    // generate a plane for test
+    vector<Point3f> plane_pts;
+    vector<float> model({1, 2, 3, 4});
+    float thr = 0.f;
+    int num = 1000;
+    vector<float> limit({5, 55, 5, 55, 0, 0});
+    generatePlane(plane_pts, model, thr, num, limit);
+
+    // get knn search result
+    Mat knn_idx;
+    int k = 10;
+    getKNNSearchResultsByKDTree(knn_idx, noArray(), plane_pts, k);
+
+    // estimate normal and curvature
+    vector<Point3f> normals;
+    vector<float> curvatures;
+    normalEstimate(normals, curvatures, plane_pts, knn_idx, k);
+
+    float theta_thr = 1.f; // degree of angle between normal of plane and normal of point
+    float curvature_thr = 0.01f; // threshold for curvature and actual curvature of the point
+    float actual_curvature = 0.f;
+
+    Point3f n1(model[0], model[1], model[2]);
+    float n1m = n1.dot(n1);
+    float total_theta = 0.f;
+    float total_diff_curvature = 0.f;
+    for (int i = 0; i < num; ++i)
+    {
+        float n12 = n1.dot(normals[i]);
+        float n2m = normals[i].dot(normals[i]);
+        float cos_theta = n12 / sqrt(n1m * n2m);
+        // accuracy problems caused by float numbers, need to be fixed
+        cos_theta = cos_theta > 1 ? 1 : cos_theta;
+        cos_theta = cos_theta < 0 ? 0 : cos_theta;
+        float theta = acos(cos_theta);
+
+        total_theta += theta;
+        total_diff_curvature += abs(curvatures[i] - actual_curvature);
+    }
+
+    float avg_theta = total_theta / (float) num;
+    ASSERT_LE(avg_theta, theta_thr);
+
+    float avg_diff_curvature = total_diff_curvature / (float) num;
+    ASSERT_LE(avg_diff_curvature, curvature_thr);
+}
+
+TEST(NormalEstimationTest, SphereNormalEstimation)
+{
+    // generate a sphere for test
+    vector<Point3f> sphere_pts;
+    vector<float> model({0, 0, 0, 1});
+    float thr = 0.f;
+    int num = 1000;
+    vector<float> limit({-1, 1, -1, 1, -1, 1});
+    generateSphere(sphere_pts, model, thr, num, limit);
+
+    // get knn search result
+    Mat knn_idx;
+    int k = 10;
+    getKNNSearchResultsByKDTree(knn_idx, noArray(), sphere_pts, k, nullptr);
+
+    // estimate normal and curvature
+    vector<Point3f> normals;
+    vector<float> curvatures;
+    normalEstimate(normals, curvatures, sphere_pts, knn_idx, k);
+
+    float theta_thr = 1.f; // degree of angle between normal of plane and normal of point
+    float curvature_thr = 0.01f; // threshold for curvature and actual curvature of the point
+    float actual_curvature = 0.f;
```

Comment:
Is it OK that actual curvature of the sphere is set to zero?

Reply:
The curvature of the sphere is not 0. Here is how I understand it: the normal estimation is made on a sufficiently small part of the surface. These nearest neighbor points are then approximated as a plane, and the smaller the curvature of these points, the more accurate the estimation of the normal will be.

Reply:
Sounds strange since a curvature (pls correct me if I'm wrong) approximates a surface in 2nd order manner.
If you have some source of that test then maybe they also contain some justification for that.

Reply:
The values estimated using PCA are approximate curvature rather than exact curvature. This method was chosen because it is faster and easier and other methods may be more accurate, which may be implemented subsequently.

![image](https://user-images.githubusercontent.com/49567307/168948901-36ead486-97b7-48c0-a050-f44459b2a097.png)

From Figure 8c and 8d in paper [Robust Segmentation for Large Volumes of Laser
Scanning Three-Dimensional Point Cloud Data](https://ieeexplore.ieee.org/document/7465797), we can see that the larger the proportion of neighborhood size in the non-planar estimation, the more inaccurate the PCA estimation results are. Therefore, I designed this part to ensure that the chosen `k` value is not too large relative to the point cloud to ensure that the accuracy of the normal estimation of the surface is also acceptable. For the PCA-based normal estimation, the estimation result is meaningless if `k` is too large. Therefore, it may be more appropriate to understand the normal estimation of the surface according to the local plane. (It seems that it is meaningless to use a non-planar to test PCA-based normal estimation. Maybe I should delete this part?)